dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -478,27 +483,232 @@ public Object put(String key, Object value) {
 
     @Override
     public Set<String> keySet() {
-        Set<String> myKeys = new HashSet<String>(super.keySet());
-        for (int i = 0; i < TOTAL; i++) {
-            if (propertyValues[i] != NOT_FOUND) {
-                myKeys.add(propertyNames[i]);
+        return new KeySet();
+    }
+    
+    Set<String> backedKeySet() {
+        return super.keySet();
+    }
+    
+    Set<Map.Entry<String, Object>> backedEntrySet() {
+        return super.entrySet();
+    }
+    
+    Collection<Object> backedValues() {
+        return super.values();
+    }
+    
+    abstract class MessageIterator<T> implements Iterator<T>{
+        Iterator<T> backedIterator;
+        int current = -1;
+        int next = -1;
+        boolean removeAllowed = false;
+        MessageIterator(Iterator<T> it) {
+            backedIterator = it;
+        }
+        @Override
+        public final boolean hasNext() {
+            if (backedIterator.hasNext()) {
+                return true;
+            }
+            if (next == TOTAL) {
+                return false;
+            }
+
+            for (int i = current+1; i < TOTAL; i++) {
```

Comment:
I guess we could have done this simpler:
```
for (next = current+1; next < TOTAL; next++) {
   if (propertyesValues[next] != NOT_FOUND) {
...
```

then you don't need to set next = i or next = TOTAL because it is already set.